### PR TITLE
Continuous pcap stream integration

### PR DESCRIPTION
**Pull request by mereacre**
Added libpcap streaming options.

Capture packets and svaes to SQLite DB or pipes to named pipes using protobuf serialization.

To Do:
 - Need to to add a queue or ring buffer to store arriving packets as there's a high drop rate.

### CODE REVIEW

File: src/recap.c
Comment (aloisklink) [inline src/recap.c:91]: Why `-t` :tea: ?

I'm mainly asking because in PR #348, I used the exact same character (`-t` for `transaction`) (great minds think alike), see https://github.com/nqminds/edgesec/blob/736e0917212ffc27e4a9e5fa25d03d9a22934102/src/recap.c#L122-L132 :)

I did try looking at `getopts_long(` so that I could use `--transaction`, but that's a glibc specific extension, so we may struggle to get it working properly on FreeBSD.

File: src/recap.c
Comment (aloisklink) [inline src/recap.c:414]:
```suggestion
int process_pcap_stream(const char *pcap_path, struct pcap_stream_context *pctx) {
```

File: src/recap.c
Comment (aloisklink) [inline src/recap.c:417]: It's always worth being a bit more verbose, especially for errors that are likely to come from user error (e.g. me accidentally writing `recap -p my-file.pcp` instead of `.pcap` or something)!

```suggestion
      log_errno("fopen failed for pcap file %s", pcap_path);
```

**Review by aloisklink — COMMENTED** [review]
Not a full review, but:

> Need to to add a queue or ring buffer to store arriving packets as there's a high drop rate.

My gut feeling is that you should also wait until the buffer is full, then do a `BEGIN TRANSACTION`, insert the entire buffer, then a `COMMIT TRANSACTION`. It will greatly increase the throughput.

---

You may also want to add my changes from PR #348 in too, since I've done a bit of refactoring, and I've added some tests for `recap`.

---

One last thing, can you think of a better name for this PR! Maybe something like `recap from libpcap stream`, or `` Add libpcap stream input to `recap` ``.

File: src/recap.c
Comment (mereacre) [reply src/recap.c:91]: It was a random selection. I can change -t to something else.

**Comment by mereacre** [comment]
There's still works to be done. Because at the moment the libcap is dropping packets when they are not precessed in time. So I will have to use the queue in the same way as I used for the header middleware.

File: src/recap.c
Comment (aloisklink) [reply src/recap.c]: > It was a random selection

I think being vegan may have given you psychic superpowers, so you read my mind and realised predicted which option char I was using :wink: :brain:

**Comment by aloisklink** [comment]
Hey @mereacre, my gut feeling is to undo https://github.com/nqminds/edgesec/pull/349/commits/bf38277ef865a3663f6a30f4ea46da067ac5d515, but instead change: https://github.com/nqminds/edgesec/blob/2582047c213c9e71cf3e170182c4e0bd922d38c5/src/recap.c#L566 to

```c
  if (sqlite_transaction_type == DEFAULT) {
    if (capture == 0) {
      sqlite_transaction_type = SINGLE_TRANSACTION;
    } else
      // disable transactions, since `recap` might run forever
      sqlite_transaction_type = DISABLED;
    }
  }
```

The way `cat my-pcap-file.pcap | recap` and `recap -p my-pcap-file.pcap` will still be super fast by default, but when you use `-n` transactions will be disabled by default.

In the future, it might be worth making an `sqlite_transaction_type = WHEN_BUFFER_FULL` too, since I'm sure that will greatly increase your insertion speed for this PR.

**Comment by aloisklink** [comment]
> I simplified the options. So now there's only one flag "-t" meaning with commit. More advanced options can be added but we need to see if there's a use for them.

It does mean that it's now disabled by default though, which I don't really like.

Ideally it would enabled when importing a `.pcap` file, and disabled by default when using a continuous stream from `libpcap`, but I guess it's not too hard to just add the `-t` flag.

**Comment by mereacre** [comment]
I agree. I will enable by default for pcap capture.

If you prefer I can enable by default the commit for pcap file and stream.

File: src/capture/middlewares/header_middleware/packet_decoder.c
Comment (aloisklink) [inline src/capture/middlewares/header_middleware/packet_decoder.c:370]: We can simplify to the following:

```suggestion
  generate_radom_uuid(cpac.id);
```

File: src/capture/middlewares/protobuf_middleware/protobuf_middleware.h
Comment (aloisklink) [inline src/capture/middlewares/protobuf_middleware/protobuf_middleware.h:20]: This variable may be changed by the function.

```suggestion
 * @param fd[in,out] The pipe file descriptor.
 * If `0`, this will be set to a new fd pointing to @p path.
```

File: src/capture/pcap_service.h
Comment (aloisklink) [inline src/capture/pcap_service.h:112]: Can you add `[in]` and `[out]` to these params? I don't know which is the input and which is the output.

(Adding `const` to the input param if possible will be nice too)

File: src/recap.c
Comment (aloisklink) [inline src/recap.c:34]: @mereacre, before you merge this, don't merge https://github.com/nqminds/edgesec/pull/116, otherwise it will cause merge conflicts (it's easy to autoformat that PR to fix the merge conflicts)

File: src/recap.c
Comment (aloisklink) [inline src/recap.c:129]: Is it worth using C99 boolean types instead?

See https://en.cppreference.com/w/c/types/boolean

File: src/recap.c
Comment (aloisklink) [inline src/recap.c:552]: Ooooh :) I like this syntax! Out of curiosity, do you know how much faster this makes your code?

File: src/recap.c
Comment (aloisklink) [inline src/recap.c:544]: You may need to improve this error. I'd even drop down this to a `log_warning()`, because I'd expect this to happen pretty often due to `SQLITE_BUSY`.

File: src/recap.c
Comment (aloisklink) [inline src/recap.c:572]: I'm surprised GCC didn't throw an error for this.

Fixes:

> ```
> /home/runner/work/edgesec/edgesec/src/recap.c:567:7: error: variable 'pc' is used uninitialized whenever 'if' condition is true [-Werror,-Wsometimes-uninitialized]
>   if ((eloop = eloop_init()) == NULL) {
>       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /home/runner/work/edgesec/edgesec/src/recap.c:608:14: note: uninitialized use occurs here
>   close_pcap(pc);
>              ^~
> /home/runner/work/edgesec/edgesec/src/recap.c:567:3: note: remove the 'if' if its condition is always false
>   if ((eloop = eloop_init()) == NULL) {
>   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /home/runner/work/edgesec/edgesec/src/recap.c:572:3: note: variable 'pc' is declared here
>   struct pcap_context *pc = NULL;
>   ^
> 1 error generated.
> ```

 
```suggestion
  struct pcap_context *pc = NULL;

  if ((eloop = eloop_init()) == NULL) {
    log_error("eloop_init fail");
    goto process_pcap_capture_fail;
  }
```

**Review by aloisklink — APPROVED** [review]
Looks good!

File: src/recap.c
Comment (mereacre) [reply src/recap.c:552]: It doesn't it make it faster in the direct sense. But it decreases the packets dropped by the libpcap. So, my first estimates is that if one saves directly into the DB libpcap drops around 30% of packets, which is not good. The committing reduces the packet drop considerable. I will have to measure it.

Comment (mereacre) [reply src/recap.c:552]: You can run the code on your PC and see many packets it drops (https://github.com/nqminds/edgesec/blob/recap-pcap/src/recap.c#L495). Eventually I will have to do similar modifications to the header middleware.

File: src/recap.c
Comment (mereacre) [reply src/recap.c:129]: I can change to bool. But I'm an old timer :) Should have used char instead.

File: src/recap.c
Comment (mereacre) [reply src/recap.c:34]: ok. I will merge this one first and then #116.

File: src/capture/pcap_service.h
Comment (mereacre) [reply src/capture/pcap_service.h:112]: Done

File: src/recap.c
Comment (mereacre) [reply src/recap.c]: Done.

File: src/recap.c
Comment (mereacre) [reply src/recap.c]: Done

**Pull request merged**
